### DISCUSSION TITLE
Use member functions for websocket client

### DISCUSSION THREAD

**Discussion by phill4y** [uNetworking/uWebSockets#1791] [Q&A]
Hi together, I am currently trying to implement a websocket client using the uWebSockets library. Unfortunately I am not able to connect the behaviour of the client to any member functions.

This is a minimal example of what I want to achieve:

```c++
#include <iostream>

#include "ClientApp.h"

using std::placeholders::_1;
using std::placeholders::_2;

struct UserData
{
    std::string url;
};

// SSL = false, isServer = true, UserData = UserData
using UWsServer = uWS::WebSocket<false, true, UserData>;

class WebSocketClient
{
public:
    WebSocketClient(const std::string& url)
        : url_(url)
    {
    }

    void run()
    {
        uWS::WebSocketClientBehavior behavior;
        behavior.open    = std::bind(&WebSocketClient::on_open, this, _1);
        behavior.message = std::bind(&WebSocketClient::on_message, this, _1, _2);
        behavior.close   = std::bind(&WebSocketClient::on_close, this, _1);

        uWS::ClientApp app(std::move(behavior));
        app.connect(url_, "protocol");
        app.run();
    }

private:
    std::string url_;
    virtual void on_message(UWsServer*, std::string_view /*message*/) {};
    virtual void on_open(UWsServer* /*ws*/) {};
    virtual void on_close(UWsServer* /*ws*/) {};
};

int main()
{
    auto test = WebSocketClient("ws://localhost:3000");
    test.run();
}

```

The compiler error I am seeing is this

```text
uWebSockets/src/MoveOnlyFunction.h:336:41: error: no matching function for call to ‘ofats::any_invocable<void()>::any_invocable(<brace-enclosed initializer list>)’
  336 |       any_invocable{std::forward<F>(f)}.swap(*this);   
```

I also tried linking the behaviour to the member methods using lambdas but couldn't get anything to work.
Any help is appreciated

**Comment by uNetworkingAB** [comment]
The client support is just a mock interface for a future implementation if that ever happens

**Comment by Sh0g0-1758** [comment]
I feel like you should mention this in the README that the library does not support creating websocket clients. I spent a considerable amount of time surfing through the code and then I see this discussion :(

**Comment by Sfinx** [comment]
Issue saved my time - found it just after trying to find out why client example silently exiting. Switching to libhv

**Comment by Sh0g0-1758** [comment]
you should try [websocketpp](https://docs.websocketpp.org/). Works like a charm.

**Reply by Sfinx** [reply]
> you should try [websocketpp](https://docs.websocketpp.org/). Works like a charm.

It is not maintainable for 4 years

**Reply by Sh0g0-1758** [reply]
still works like a charm :)
